web/src/installers/index.ts

```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
 export class Installer {
 
   public id: string;
-  public kubernetes: KubernetesConfig;
-  public weave: WeaveConfig;
-  public rook: RookConfig;
-  public contour: ContourConfig;
-  public registry: RegistryConfig;
-  public prometheus: PrometheusConfig;
-  public kotsadm: KotsadmConfig;
+  public spec: InstallerSpec;
 
   constructor(
     public readonly teamID?: string,
   ) {
-    this.kubernetes = { version: "" };
-    this.weave = { version: "" };
-    this.rook = { version: "" };
-    this.contour = { version: "" };
-    this.registry = { version: "" };
-    this.prometheus = { version: "" };
-    this.kotsadm = { version: "", applicationSlug: "" };
+    this.spec = {
+      kubernetes: { version: "" },
+    };
   }
 
+  public clone(): Installer {
+    const i = new Installer(this.teamID);
+
+    i.id = this.id;
+    i.spec = _.cloneDeep(this.spec);
+
+    return i;
+  }
+
+  // Going forward new fields are automatically included in the hash after being sorted
+  // alphabetically but the arbitrary order must be preserved for legacy fields.
   public hash(): string {
     const h = crypto.createHash('sha256');
 
-    if (this.kubernetes && this.kubernetes.version) {
-      h.update(`kubernetes_version=${this.kubernetes.version}`);
-    }
-    if (this.weave && this.weave.version) {
-      h.update(`weave_version=${this.weave.version}`);
+    if (this.spec.kubernetes && this.spec.kubernetes.version) {
+      h.update(`kubernetes_version=${this.spec.kubernetes.version}`);
     }
-    if (this.rook && this.rook.version) {
-      h.update(`rook_version=${this.rook.version}`);
+    if (this.spec.weave && this.spec.weave.version) {
+      h.update(`weave_version=${this.spec.weave.version}`);
     }
-    if (this.contour && this.contour.version) {
-      h.update(`contour_version=${this.contour.version}`);
+    if (this.spec.rook && this.spec.rook.version) {
+      h.update(`rook_version=${this.spec.rook.version}`);
     }
-    if (this.registry && this.registry.version) {
-      h.update(`registry_version=${this.registry.version}`);
+    if (this.spec.contour && this.spec.contour.version) {
+      h.update(`contour_version=${this.spec.contour.version}`);
     }
-    if (this.prometheus && this.prometheus.version) {
-      h.update(`prometheus_version=${this.prometheus.version}`);
+    if (this.spec.registry && this.spec.registry.version) {
+      h.update(`registry_version=${this.spec.registry.version}`);
```

Comment:
looks like you reverted some of my prometheus stuff

Reply:
Prometheus is included in the non-legacy hash logic below